lib/enmeshed/connector.rb

```diff
@@ -0,0 +1,179 @@
+
+    def self.parse_relationships(response)
+      parse_result(response).map {|relationship_json| Relationship.new(relationship_json) }
+    end
+    private_class_method :parse_relationships
+
+    def self.relationship_template(nbp_uid)
+      {
+        maxNumberOfAllocations: 1,
+        expiresAt: TEMPLATE_VALIDITY_PERIOD.from_now,
+        content: {
+          metadata: {nbp_uid:},
+          '@type': 'RelationshipTemplateContent',
+          onNewRelationship: {
+            '@type': 'Request',
+            items: [
+              {
+                '@type': 'ShareAttributeRequestItem',
+                mustBeAccepted: true,
+                attribute: display_name_attribute,
+                sourceAttributeId: display_name_id,
+              },
+            ] + required_attributes_template,
+          },
+        },
+      }
+    end
+    private_class_method :relationship_template
+
+    def self.required_attributes_template
+      REQUIRED_ATTRIBUTES.map do |attr|
+        {
+          '@type': 'ReadAttributeRequestItem',
+          mustBeAccepted: true,
+          query: {
+            '@type': 'IdentityAttributeQuery',
+            valueType: attr,
+          },
+        }
+      end
+    end
+    private_class_method :required_attributes_template
+
+    def self.display_name_attribute
+      {
+        '@type': 'IdentityAttribute',
+        owner: enmeshed_address,
+        value: {
+          '@type': 'DisplayName',
+          value: DISPLAY_NAME,
+        },
+      }
+    end
+    private_class_method :display_name_attribute
+
+    def self.conn
+      @conn ||= init_conn
+    end
+    private_class_method :conn
+
+    def self.init_conn
+      if User.omniauth_providers.exclude?(:nbp) || CONNECTOR_URL.nil? || API_KEY.nil? || DISPLAY_NAME.nil?
+        raise ConnectorError.new('NBP provider or enmeshed connector not configured as expected')
+      end
+
+      Faraday.new(CONNECTOR_URL, headers:)
+    end
+    private_class_method :init_conn
+
+    def self.enmeshed_address
+      return @enmeshed_address if @enmeshed_address.present?
+
+      identity = parse_result conn.get('/api/v2/Account/IdentityInfo')
+      @enmeshed_address = identity[:address]
+    end
+    private_class_method :enmeshed_address
+
+    def self.display_name_id
+      @display_name_id ||= existing_display_name_attr || create_display_name_attr
+    end
+    private_class_method :display_name_id
+
+    def self.create_display_name_attr
+      parse_result(conn.post('/api/v2/Attributes') do |req|
+        req.body = {content: display_name_attribute}.to_json
+      end)[:id]
+    end
+    private_class_method :create_display_name_attr
+
+    def self.existing_display_name_attr
+      parse_result(conn.get('/api/v2/Attributes') do |req|
+        req.params['content.@type'] = 'IdentityAttribute'
+        req.params['content.owner'] = enmeshed_address
+        req.params['content.value.@type'] = 'DisplayName'
+      end).first&.dig(:id)
```

Comment:
This should be obsolete now. Assuming that there is only one match is okay since we don't really care which display name attribute we use as long as it has the correct values.

Reply:
Yes, that's fine now, thanks!